packages/synthetics-sdk-broken-links/src/broken_links.ts

```diff
@@ -12,24 +12,187 @@
+  // START - to resolve warnings while under development
+  page;
+  link;
+  options;
+  // END - to resolve warnings  while under development
+
+  // PSEUDOCODE
+
+  // determine expected_status_code—determined here rather than `navigate(...)`
+  // since it will be used later when creating
+  // `BrokenLinksResultV1_SyntheticLinkResult` Object
+
+  // call `navigate(...)`
+  /**
+   * const {
+   *    response: responseOrError,
+   *    passed,
+   *    testing_only,
+   *    link_start_time,
+   *    link_end_time,
+   * } = await navigate(page, link, expected_status_code, options);
+   */
+
+  // Error handling based on `responseOrError` & passed:
+  //    if      `responseOrError` is type `Error` set `error_type` and
+  //            `error_message` in `SyntheticLinkResult` accordingly
+  //    else if `passed === false`. Set error `SYNTHETIC_INCORRECT_STATUS_CODE`
+
+  // return `SynheticLinkResult` with all calculated information
+  return {} as BrokenLinksResultV1_SyntheticLinkResult;
+}
+
+/**
+ * Navigates to a target URL with retries and timeout handling.
+ *
+ * @param page - The Puppeteer Page instance.
+ * @param link - The LinkIntermediate containing the target URL.
+ * @param expected_status_code - The expected HTTP status code.
+ * @param options - The options for navigation and retries.
+ * @returns Information about navigation attempt:
+ *   - `response`: HTTP response or error if navigation fails, or null.
+ *   - `passed`: Boolean indicating if navigation passed per status code.
+ *   - `retriesRemaining`: Remaining retries after attempt. (for testing)
+ *   - `link_start_time`: Start time of navigation attempt.
+ *   - `link_end_time`: End time of navigation attempt.
+ */
+export async function navigate(
+  page: Page,
+  link: LinkIntermediate,
+  options: BrokenLinksResultV1_BrokenLinkCheckerOptions,
+  expected_status_code: ResponseStatusCode = {
+    status_class: ResponseStatusCode_StatusClass.STATUS_CLASS_2XX,
+  }
+): Promise<{
+  response: HTTPResponse | Error | null;
+  passed: boolean;
+  retriesRemaining: number;
+  link_start_time: string;
+  link_end_time: string;
+}> {
+  let link_start_time = '';
+  let link_end_time = '';
+  let retriesRemaining = options.max_retries!;
+  const per_link_timeout_millis =
+    options.per_link_options[options.origin_url]?.link_timeout_millis ||
+    options.link_timeout_millis!;
+
+  let response: HTTPResponse | Error | null = null;
+
+  let passed = false;
+  let used_retry = false;
+  while (retriesRemaining > 0 && !passed) {
+    retriesRemaining--;
+    link_start_time = new Date().toISOString();
+    response = await fetchLink(page, link.target_url, per_link_timeout_millis);
+    link_end_time = new Date().toISOString();
+
+    // prevents errors caused by navigating from one url to same url with a
+    // different anchor part (normally returns null)
+    // e.g. mywebsite.com#heading1 --> mywebsite.com#heading2
+    if (response === null && !used_retry) {
```

Comment:
Two observations:
- `used_retry` makes the retry behavior hard to predict/understand
- the way we handle this edge case doesn't seem to provide much value

So it seems like there is state which persists in puppeteer between calls to `navigate`, and when there are consecutive links to the same URL, puppeteer doesn't request the page again and gives us an empty response (which we workaround by clearing the page and retrying). Did I get that right?

Reply:
One thing to note is that retrying to the same URL will cause puppeteer to retry the page. The HTTP response is only null if the #anchor changes.
Which is why weonly need logic in if statement to execute at most once